lib/membrane/timestamp_queue.ex

```diff
@@ -0,0 +1,389 @@
+  from `pad_ref`.
+  """
+  @spec wait_on_pad(t(), Pad.ref()) :: t()
+  def wait_on_pad(%__MODULE__{} = timestamp_queue, pad_ref) do
+    timestamp_queue
+    |> Map.update!(:waiting_on_buffer_from, &MapSet.put(&1, pad_ref))
+  end
+
+  @doc """
+  Pushes a buffer associated with a specified pad to the queue.
+
+  Returns a suggested actions list and the updated queue.
+
+  If amount of buffers associated with specified pad in the queue just exceded
+  `pause_demand_boundary`, the suggested actions list contains `t:Action.pause_auto_demand()`
+  action, otherwise it is equal an empty list.
+
+  Buffers pushed to the queue must have a non-`nil` `dts` or `pts`.
+  """
+  @spec push_buffer(t(), Pad.ref(), Buffer.t()) :: {[Action.pause_auto_demand()], t()}
+  def push_buffer(_timestamp_queue, pad_ref, %Buffer{dts: nil, pts: nil} = buffer) do
+    raise """
+    #{inspect(__MODULE__)} accepts only buffers whose dts or pts is not nil, but it received\n#{inspect(buffer, pretty: true)}
+    from pad #{inspect(pad_ref)}
+    """
+  end
+
+  def push_buffer(%__MODULE__{} = timestamp_queue, pad_ref, buffer) do
+    buffer_size = timestamp_queue.metric.buffers_size([buffer])
+
+    timestamp_queue
+    |> Map.update!(:waiting_on_buffer_from, &MapSet.delete(&1, pad_ref))
+    |> push_item(pad_ref, {:buffer, buffer})
+    |> get_and_update_in([:pad_queues, pad_ref], fn pad_queue ->
+      pad_queue
+      |> Map.merge(%{
+        buffers_size: pad_queue.buffers_size + buffer_size,
+        buffers_number: pad_queue.buffers_number + 1
+      })
+      |> Map.update!(:timestamp_offset, fn
+        nil -> timestamp_queue.current_queue_time - (buffer.dts || buffer.pts)
+        valid_offset -> valid_offset
+      end)
+      |> Map.update!(:use_pts?, fn
+        nil -> buffer.dts == nil
+        valid_boolean -> valid_boolean
+      end)
+      |> check_timestamps_consistency!(buffer, pad_ref)
+      |> actions_after_pushing_buffer(pad_ref, timestamp_queue.pause_demand_boundary)
+    end)
+  end
+
+  defp check_timestamps_consistency!(pad_queue, buffer, pad_ref) do
+    if not pad_queue.use_pts? and buffer.dts == nil do
+      raise """
+      Buffer #{inspect(buffer, pretty: true)} from pad #{inspect(pad_ref)} has nil dts, while \
+      the first buffer from this pad had valid integer there. If the first buffer from a pad has \
+      dts different from nil, all later buffers from this pad must meet this property.
+      """
+    end
+
+    buffer_timestamp = if pad_queue.use_pts?, do: buffer.pts, else: buffer.dts
+    max_timestamp = pad_queue.max_timestamp_on_qex
+
+    if is_integer(max_timestamp) and max_timestamp > buffer_timestamp do
+      raise """
+      Buffer #{inspect(buffer, pretty: true)} from pad #{inspect(pad_ref)} has timestamp equal \
+      #{inspect(buffer_timestamp)}, but previous buffer pushed on queue from this pad had timestamp \
+      equal #{inspect(max_timestamp)}. Buffers from a single pad must have non-decreasing timestamps.
```

Comment:
I can refactor error message, but simply changing `timestamp` to `PTS` or `DTS` will make this log untrue, since having buffers with monotonic DTS and non-monotonic PTS is totally ok

Reply:
> having buffers with monotonic DTS and non-monotonic PTS is totally ok

If the queue relies on PTS and they are not monotonic, the queue can't handle this case no matter what DTS are, so I think the log would be true. Such a situation would probably mean a bug anyway - I believe PTS and DTS should either be always set or never set.

Reply:
If buffers from a specific pad contain DTS, queue will look only at DTS and won't look at PTS